File: 120-Triangle/120-Triangle.cpp
// Last updated: 6/10/2026, 7:26:46 PM
class Solution {
public:
    int minimumTotal(vector<vector<int>>& a) {
        for(int i=1 ; i<a.size() ; i++) {
            for(int j=0 ; j<a[i].size() ; j++){ 
                if(j+1 == a[i].size())
                    a[i][j] += a[i-1][j-1];
                else if(j == 0)
                    a[i][j] += a[i-1][j];
                else
                    a[i][j] += min(a[i-1][j-1] , a[i-1][j]);
            }
        }
        return (int)(*min_element(a.back().begin() , a.back().end()));
    }
};

File: 120-Triangle/120-Triangle.py
# Last updated: 6/10/2026, 7:26:41 PM
class Solution(object):
    def minimumTotal(self, triangle):
        n = len(triangle)
    
        # Each node will be identified as (r, c)
        graph = defaultdict(list)
        
        # Build DAG edges
        for r in range(n - 1):
            for c in range(len(triangle[r])):
                u = (r, c)
                v1 = (r + 1, c)
                v2 = (r + 1, c + 1)
                graph[u].append((v1, triangle[r + 1][c]))
                graph[u].append((v2, triangle[r + 1][c + 1]))
        
        # Distance dictionary, start from top
        dist = {(0, 0): triangle[0][0]}
        
        # Topological order is simply row by row
        for r in range(n):
            for c in range(len(triangle[r])):
                if (r, c) not in dist:
                    continue
                for v, w in graph[(r, c)]:
                    new_dist = dist[(r, c)] + w
                    if v not in dist or new_dist < dist[v]:
                        dist[v] = new_dist
        
        # Answer = min dist among last row nodes
        last_row = [(n - 1, c) for c in range(n)]
        return min(dist[v] for v in last_row)

